Fix Host Commands for STM32H7 chips

niedzwiecki-dawid opened the pull request:
There are several changes to make Host Commands working on STM32H7 chips:
- fix clock domain support
- fix CS interrupt configuration
- handle underrun event
- restart SPI module after every transaction
- add possibility to place common buffers in nocache section

niedzwiecki-dawid committed:
mgmt: ec_host_cmd: add support for all dma versions

Update macro used to get fifo threshold from DTS, to support all
STM32 DMA versions.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid committed:
mgmt: ec_host_cmd: stm32_spi: fix clock domain support

Fix support for clock domains. Create define that identify if
clock domain was selected in any of the st_stm32_spi_host_cmd nodes.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid committed:
mgmt: ec_host_cmd: stm32_spi: fix cs interrupt configuration

Move enabling CS interrupt at the end of the backend configuration,
when everything is ready.

It prevents handling the interrupt, when something is not prepared e.g.
DMA is not configured.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid committed:
mgmt: ec_host_cmd: stm32_spi: handle underrun for stm32h7 chips

STM32H7 chips have additional configuration regarding underrun event.
Configure the underrun behaviour correctly to use the underrun pattern
and clear the underrun bit before sending new data.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid committed:
mgmt: ec_host_cmd: stm32_spi: fix rx reload for STM32H7 chips

Disable and enable SPI module before every transaction for STM32H7
chips. It is a recommended way in the STM32 RM. In another case, a first
byte of a transaction is always 0x00.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid committed:
mgmt: ec_host_cmd: add support for nocache buffers

Some backends use DMA. Usually, DMA doesn't work correctly, when memory
to transfer is cached. Add a config to place the common buffers in the
nocache section.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid committed:
mgmt: ec_host_cmd: stm32_spi: fix clang-format

Use clang-format for the ec_host_cmd_backend_spi_stm32.c file.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

duda-patryk reviewed (commented):
Looks good to me. I only found two things that can be improved.

erwango commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_spi_stm32.c:
```diff
@@ -107,6 +107,12 @@ BUILD_ASSERT(DT_NODE_HAS_COMPAT_STATUS(DT_CHOSEN(zephyr_host_cmd_spi_backend),
 #define STM32_DMA_FEATURES_ID(id, dir) DT_DMAS_CELL_BY_NAME(id, dir, features)
 #endif
 
+#if DT_CLOCKS_HAS_IDX(DT_CHOSEN(zephyr_host_cmd_spi_backend), 1)
```
I hope `STM32_DOMAIN_CLOCK_SUPPORT` should work.
Though, maybe removing the whole `STM32_SPI_DOMAIN_CLOCK_SUPPORT/STM32_EC_HOST_CMD_SPI_DOMAIN_CLOCK_SUPPORT` would be a better idea. It doesn't save much compared to the induced "complexity".

niedzwiecki-dawid replied:
Yes, `STM32_DOMAIN_CLOCK_SUPPORT` doesn't seem correct. It is not that complicated and it is what other drivers do, so I would stick to the current version.

duda-patryk replied:
Sounds good.

niedzwiecki-dawid committed:
include/devicetree: dma: add DT_DMAS_CELL_BY_NAME_OR macro

Add the DT_DMAS_CELL_BY_NAME_OR macro. It is almost the same as
DT_DMAS_CELL_BY_NAME, but it uses DT_PHA_BY_NAME_OR instead of
DT_PHA_BY_NAME.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_spi_stm32.c:
```diff
@@ -100,6 +100,13 @@ BUILD_ASSERT(DT_NODE_HAS_COMPAT_STATUS(DT_CHOSEN(zephyr_host_cmd_spi_backend),
 #define EC_HOST_CMD_ST_STM32_FIFO
 #endif /* st_stm32_spi_fifo */
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma_v2) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma_v2bis) || \
```
`DT_DMAS_CELL_BY_NAME_OR` added

erwango commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_spi_stm32.c:
```diff
@@ -195,8 +197,7 @@ static int prepare_rx(struct ec_host_cmd_spi_ctx *hc_spi);
 	},                                                                                         \
-	.fifo_threshold =                                                                          \
-		STM32_DMA_FEATURES_FIFO_THRESHOLD(DT_DMAS_CELL_BY_NAME(id, dir, features)),
+	.fifo_threshold = STM32_DMA_FEATURES_FIFO_THRESHOLD(STM32_DMA_FEATURES_ID(id, dir)),
```
The new macro is nice, but what does it bring compare to existing definition ?
```
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma_v1)
#define STM32_DMA_FEATURES_FIFO_THRESHOLD(features)	(features & 0x3)
#else
#define STM32_DMA_FEATURES_FIFO_THRESHOLD(features)	0
#endif
```

erwango replied:
Ok, got it.
Actually those (`STM32_DMA_FEATURES`) should be `_INST_` variants and we should have provided "non compat implicit" versions as you need. But you're proposal is fine as well indeed

erwango commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_spi_stm32.c:
```diff
@@ -536,13 +536,28 @@ static int prepare_rx(struct ec_host_cmd_spi_ctx *hc_spi)
 
 	hc_spi->prepare_rx_later = 0;
+
+#ifdef EC_HOST_CMD_ST_STM32H7
```
Note: `st,stm32h7-spi` is also used in other series like H5, U5, WBA.
I didn't check the RM for those, but you might want to check that tis behavior is related to the IP rather than the series.

niedzwiecki-dawid replied:
Thanks for the info. I've checked and HAL provides used functions inside `#ifdef EC_HOST_CMD_ST_STM32H7` also for H5, U5 and WBA chips.
